Bug 1997226: Fix enabling PROXY protocol on an upgraded cluster

Miciah opened the pull request:
#### `setDefaultPublishingStrategy`: Reformat with `switch`

Refactor the update logic in `setDefaultPublishingStrategy`.  Also, fix `setDefaultPublishingStrategy` to return true if the scope changed.  (Nothing uses this return value, so this change is only for correctness.)

* `pkg/operator/controller/ingress/controller.go` (`setDefaultPublishingStrategy`): Use a switch statement for the update logic so that the logic only looks at parameters related to the selected endpoint publishing strategy type.


#### `setDefaultPublishingStrategy`: Fix PROXY protocol

Fix the update logic in `setDefaultPublishingStrategy` so that updates are properly handled when `status.endpointPublishingStrategy.hostNetwork` or `status.endpointPublishingStrategy.nodePort` is null.

Before OpenShift 4.8, the IngressController API did not have any fields under the `status.endpointPublishingStrategy.hostNetwork` and `status.endpointPublishingStrategy.nodePort` fields.  As result, these fields could be null even if the `spec.endpointPublishingStrategy.type` field was set to "HostNetwork" or "NodePortService".

OpenShift 4.8 added `status.endpointPublishingStrategy.hostNetwork.protocol` and `status.endpointPublishingStrategy.nodePort.protocol` fields, and the operator now sets default values for these fields when the operator admits or re-admits an ingresscontroller that specifies the "HostNetwork" or "NodePortService" strategy type, respectively.

However, a cluster that was upgraded from a version of OpenShift before 4.8 could have an already admitted ingresscontroller with null values for `status.endpointPublishingStrategy.hostNetwork` and `status.endpointPublishingStrategy.nodePort` even when ingresscontroller specifies the "HostNetwork" or "NodePortService" strategy type.

In this case, the operator ignored updates to the `spec.endpointPublishingStrategy.hostNetwork.protocol` or `spec.endpointPublishingStrategy.nodePort.protocol` fields.

This PR fixes the update logic so that it correctly updates the `status.endpointPublishingStrategy.hostNetwork.protocol` or `status.endpointPublishingStrategy.nodePort.protocol` field when `status.endpointPublishingStrategy.hostNetwork` or `status.endpointPublishingStrategy.nodePort` is null, the `spec.endpointPublishingStrategy.hostNetwork.protocol` or `spec.endpointPublishingStrategy.nodePort.protocol` field is set, and the strategy type is "HostNetwork" or "NodePortService", respectively.

* `pkg/operator/controller/ingress/controller.go` (`setDefaultPublishingStrategy`): Fix logic to properly handle null values for `status.endpointPublishingStrategy.hostNetwork` or `status.endpointPublishingStrategy.nodePort`.


---

@miheer, does this look like it will resolve the issue?

miheer commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -396,40 +396,47 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, infraConfig
+			// If the ProviderParameters field does not exist for spec or status,
+			// just propagate (or remove) ProviderParameters in its entirety
+			// (as long as GCP parameters are specified one way or the other).
+			if specLB.ProviderParameters == nil && statusLB.ProviderParameters != nil && statusLB.ProviderParameters.GCP != nil ||
```
@Miciah just to understand... Why do we conditions specific for GCP only ?

Miciah replied:
GCP is the only platform where the provider parameters can be changed without deleting and recreating the service load-balancer.  In particular, GCP has the "global client access" option that can be turned on and off.  The only other platform with provider parameters is AWS, which has an option to specify whether a Classic ELB or an NLB should be provisioned, and changing that option requires deleting and recreating the service load-balancer.

Miciah commented:
Rebased to resolve the conflict from #582, and also fixed a tiny, non-consequential issue related to the return value of `setDefaultPublishingStrategy` when the scope has been change.

Miciah commented:
Must-gather failed (couldn't connect to the API).
/test e2e-aws-single-node

miheer commented:
@Miciah  we don't need unit tests for this ? Or may we can do it in a followup PR ?

miheer commented:
/lgtm

miheer commented:
/approve

candita commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -396,44 +396,63 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, infraConfig
+		}
+		statusNP := ic.Status.EndpointPublishingStrategy.NodePort
+		specNP := effectiveStrategy.NodePort
+		if specNP != nil && specNP.Protocol != statusNP.Protocol {
```
Is it possible to try to set NodePort to nil?

candita commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -396,44 +396,63 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, infraConfig
+		statusNP := ic.Status.EndpointPublishingStrategy.NodePort
+		specNP := effectiveStrategy.NodePort
+		if specNP != nil && specNP.Protocol != statusNP.Protocol {
+			statusNP.Protocol = specNP.Protocol
```
```suggestion
			ic.Status.EndpointPublishingStrategy.NodePort.Protocol = specNP.Protocol
```

candita commented on pkg/operator/controller/ingress/controller.go:
```diff
@@ -396,44 +396,63 @@ func setDefaultPublishingStrategy(ic *operatorv1.IngressController, infraConfig
+		statusHN := ic.Status.EndpointPublishingStrategy.HostNetwork
+		specHN := effectiveStrategy.HostNetwork
+		if specHN != nil && specHN.Protocol != statusHN.Protocol {
+			statusHN.Protocol = specHN.Protocol
```
```suggestion
			ic.Status.EndpointPublishingStrategy.HostNetwork.Protocol = specHN.Protocol
```

Miciah replied:
`ic.Status.EndpointPublishingStrategy.HostNetwork` is a pointer, so the following:
```go
statusHN := ic.Status.EndpointPublishingStrategy.HostNetwork
statusHN.Protocol = specHN.Protocol
```
is equivalent to the following:
```go
ic.Status.EndpointPublishingStrategy.HostNetwork.Protocol = specHN.Protocol
```
I'll rustle up some unit tests.

candita replied:
It would be nice to see the same style throughout this `switch` then.

candita commented:
@Miciah I made some comments.  Can you add a unit test for this?

Miciah commented:
/cherry-pick release-4.9
/cherry-pick release-4.8
